File: src/core/session.ts
import { randomUUID } from 'node:crypto';
import type { Session, Intensity, SessionOutcome, Turn, SkillScores, BehaviorTag, Quote } from '../types.js';
import { SCHEMA_VERSION } from '../types.js';

export function createSession(problemDescription: string, intensity: Intensity): Session {
  return {
    schemaVersion: SCHEMA_VERSION,
    id: randomUUID(),
    timestamp: new Date().toISOString(),
    endTimestamp: '',
    problemDescription,
    intensity,
    outcome: 'abandoned',
    questionsToRootCause: 0,
    codeFiles: [],
    transcript: [],
    skillScores: { assumptionChecking: 5, evidenceGathering: 5, rootCauseSpeed: 5 },
    behaviorTags: [],
  };
}

export function addTurn(
  session: Session,
  question: string,
  response: string,
  model: string,
  behaviorTags: string[] = [],
  quoteUsed: Quote | null = null,
): Turn {
  const turn: Turn = {
    turnNumber: session.transcript.length + 1,
    question,
    response,
    model,
    quoteUsed,
    behaviorTags,
    timestamp: new Date().toISOString(),
  };
  session.transcript.push(turn);
  session.questionsToRootCause = session.transcript.length;
  return turn;
}

export function finalizeSession(
  session: Session,
  outcome: SessionOutcome,
  skillScores?: SkillScores,
  behaviorTags?: BehaviorTag[],
): void {
  session.endTimestamp = new Date().toISOString();
  session.outcome = outcome;
  if (skillScores) session.skillScores = skillScores;
  if (behaviorTags) session.behaviorTags = behaviorTags;
}

export function addCodeFile(session: Session, filePath: string): void {
  if (!session.codeFiles.includes(filePath)) {
    session.codeFiles.push(filePath);
  }
}

File: nietzschian_debugger/core/session.py
"""Session creation and management."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Optional

from ..types import (
    BehaviorTag,
    Intensity,
    Quote,
    SCHEMA_VERSION,
    Session,
    SessionOutcome,
    SkillScores,
    Turn,
)


def create_session(problem_description: str, intensity: Intensity) -> Session:
    """Create a new debugging session."""
    return Session(
        schema_version=SCHEMA_VERSION,
        id=str(uuid.uuid4()),
        timestamp=datetime.now(timezone.utc).isoformat(),
        end_timestamp="",
        problem_description=problem_description,
        intensity=intensity,
        outcome="abandoned",
        questions_to_root_cause=0,
        code_files=[],
        transcript=[],
        skill_scores=SkillScores(
            assumption_checking=5,
            evidence_gathering=5,
            root_cause_speed=5,
        ),
        behavior_tags=[],
    )


def add_turn(
    session: Session,
    question: str,
    response: str,
    model: str,
    behavior_tags: Optional[list[str]] = None,
    quote_used: Optional[Quote] = None,
) -> Turn:
    """Add a turn to the session transcript."""
    turn = Turn(
        turn_number=len(session.transcript) + 1,
        question=question,
        response=response,
        model=model,
        quote_used=quote_used,
        behavior_tags=behavior_tags or [],
        timestamp=datetime.now(timezone.utc).isoformat(),
    )
    session.transcript.append(turn)
    session.questions_to_root_cause = len(session.transcript)
    return turn


def finalize_session(
    session: Session,
    outcome: SessionOutcome,
    skill_scores: Optional[SkillScores] = None,
    behavior_tags: Optional[list[BehaviorTag]] = None,
) -> None:
    """Finalize a session with outcome and scores."""
    session.end_timestamp = datetime.now(timezone.utc).isoformat()
    session.outcome = outcome
    if skill_scores is not None:
        session.skill_scores = skill_scores
    if behavior_tags is not None:
        session.behavior_tags = behavior_tags


def add_code_file(session: Session, file_path: str) -> None:
    """Add a code file to the session if not already present."""
    if file_path not in session.code_files:
        session.code_files.append(file_path)
